resource/kind.go

```diff
@@ -0,0 +1,319 @@
+package resource
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+)
+
+var (
+	_ Kind    = &UntypedKind{}
+	_ Object2 = &UntypedObject{}
+)
+
+// Object2 implements kubernetes' runtime.Object and meta/v1.Object, as well as some additional methods useful for the app-sdk
+type Object2 interface {
+	runtime.Object
+	metav1.Object
+	GetSpec() any
+	GetSubresources() map[string]any
+	GetStaticMetadata() StaticMetadata
+	SetStaticMetadata(metadata StaticMetadata)
+}
+
+// MultiVersionKind is a collection of Kinds which have the same Group and Kind, but different versions.
+// This may allow us to include conversion as a component of the interface.
+// TODO: not sure if necessary
+type MultiVersionKind interface {
+	Group() string
+	Kind() string
+	Versions() []Kind
+}
+
+// Group is a collection of Kinds which share a Group and Version, allowing them to be put under the same
+// routing in an API server.
+// TODO: not sure if necessary
+type Group interface {
+	Group() string
+	Version() string
+	Kinds() []Kind
+}
+
+// Kind is an interface representing a kubernetes-compatible Kind, which contains information about the Kind's
+// Group, Version, Kind, Plural, and Scope, as well as allowing for creation of a zero-value version of the Kind as an
+// Object (which may contain default values for fields as decided by the implementation).
+// Additionally, Kind contains the logic for reading wire bytes into an Object instance of the Kind,
+// or serializing an Object instance of the Kind into wire bytes.
+// In this sense, Kind combines schema information and a serializer into one interface used to interact with the kind.
+type Kind interface {
+	//Schema <- can't compose because we need Object2
+	// TODO: just update Object instead of using Object2?
```

Comment:
What's the main reason behind not updating Object? Is it just because of the breaking changes?

Reply:
I intend to replace `Object` once we think this is a good path, initially I was unsure if we'd want to use the same name, and also `Object` is used in a lot of places and I wanted to avoid having to deal with everywhere it was used initially--the intent is to replace `Object` before this gets merged, unless we can think of some better naming.